src/cart/crt_swim.c

```diff
@@ -1366,10 +1366,9 @@ int crt_swim_rank_add(struct crt_grp_priv *grp_priv, d_rank_t rank)
 			if (cst == NULL)
 				D_GOTO(out_unlock, rc = -DER_NOMEM);
 		}
-		id = (swim_id_t)rank;
-		cst->cst_id = id;
+		cst->cst_id = rank;
 		cst->cst_state.sms_incarnation = 0;
-		cst->cst_state.sms_status = SWIM_MEMBER_INACTIVE;
+		cst->cst_state.sms_status = SWIM_MEMBER_ALIVE;
```

Comment:
Hmm.. this starts all members in alive state. Is this going to be an issue at the startup? I remember vaguely there was some issue Dmitry was trying to avoid before which is why he introduced inactive state.

From looking through code the only other place where member state can change to inactive is in crt_swim_suspend_all() call, which is not used.

If based on your analysis we dont need inactive state anymore, then probably should also cleanup the code here to make state machine easier to read:)

Reply:
@frostedcmos, see the last item in my commit message. INACTIVE was introduced to mitigate the "join" race, that is:

1. Engine X sends a JoinReq to the MS.
2. The MS sends a JoinResp to X, and informs everyone, including engine Y, that X is part of the system.
3. Y sends a SWIM_PING to X.
4. X receives the SWIM_PING.
5. X receives the JoinResp and sets its rank.
6. X receives its initial group membership.

At that time (2018?), swim used one-way RPCs, and had some technical issues with sending a SWIM_PING response back to Y. Hence, Y may suspect X, and in theory might cause X to be considered as dead. Moreover, swim didn't have a mechanism for X to resurrect from the dead.

With normal, two-way RPCs swim uses today, X can reply to the SWIM_PING (or a SWIM_IREQ), saying that "I'm alive but not fully initialized yet." Y then considers the direct ping (or a indirect ping) to be successful. Therefore, AFAICS, we no longer need to put new members in the INACTIVE state. (Of course, maybe there are some other considerations I'm not aware of.)

I believe there's an unrelated use of INACTIVE as you pointed out. Since I don't have a good understanding for that use (related to the network "glitch" or unavailabilty stuff?), could we leave the code there for the moment?

Reply:
Thanks for the explanation, and yes, I agree, we can leave for now that code.

Reply:
Thanks for the understanding.